src/Tools/DynamoInstallDetective/ProductLookUp.cs

```diff
@@ -9,6 +9,118 @@
+        // Dictionary<ProductCode, (DisplayName, InstallPath)>
+        private static Dictionary<string, Tuple<string, string>> cachedRecords;
+        private static bool cacheEnabled = false;
+
+        private static readonly object mutex = new object();
+
+        public static RegistryKey OpenKey(string key)
+        {
+            var regKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64);
+            return regKey.OpenSubKey(key);
+        }
+        public static string GetInstallLocation(RegistryKey key)
+        {
+            if (key != null)
+                return key.GetValue("InstallLocation") as string;
+
+            return string.Empty;
+        }
+
+        public static string GetDisplayName(RegistryKey key)
+        {
+            if (key != null)
+                return key.GetValue("DisplayName") as string;
+
+            return string.Empty;
+        }
+
+        // Returns all the products registered under "SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\" that have a valid DisplayName and an InstallLocation.
+        public static Dictionary<string, Tuple<string, string>> GetInstalledProducts()
+        {
+            lock (mutex)
+            {
+                if (cacheEnabled && cachedRecords != null)
+                {
+                    return cachedRecords;
+                }
+            }
+
+            var productInfo = new Dictionary<string, Tuple<string, string>>();
+
+            var key = OpenKey(REG_KEY64);
+            foreach(string s in key.GetSubKeyNames())
+            {
+                try
+                {
+                    var prodKey = key.OpenSubKey(s);
+                    var displayName = GetDisplayName(prodKey);
+                    var installLocation = GetInstallLocation(prodKey);
+                    if (displayName.Length > 0 && installLocation.Length > 0)
+                    {
+                        productInfo.Add(s, Tuple.Create(displayName, installLocation));
+                    }
+                }
+                catch (Exception)
+                { }
+            }
+
+            lock (mutex)
+            {
+                if (cacheEnabled)
+                {
+                    cachedRecords = new Dictionary<string, Tuple<string, string>>(productInfo);
+                }
+            }
+            return productInfo;
+        }
+
+        public static IDisposable StartCache()
+        {
+            return new RegistryCacher();
+        }
+
+        public static void EnableCache()
+        {
+            lock (mutex)
+            {
+                cacheEnabled = true;
+            }
+        }
+
+        public static void DisableCache()
+        {
+            lock (mutex)
+            {
+                cacheEnabled = false;
+                cachedRecords.Clear();
+                cachedRecords = null;
+            }
+        }
+    }
```

Comment:
I think all methods of `RegUtils` can also be made internal as the class itself is internal and does not derive from any other classes or interfaces.

Reply:
If the Class is internal then it means, regardless of the method accessor, methods will not be accessible outside the current assembly. 
Having public specifier for certain methods in an internal class only helps to clarify the intention that those methods would be safe to use if the class was ever to be made public.

Reply:
I think it could be confusing to make the class internal and the methods public. It's more explicit to make the methods internal as well, but that's just me.

Reply:
To me it sounds a bit overkill to make all members internal if a class is specified as internal.
I guess if we consider this as a coding convention I could get behind it. 
What do you think @mjkkirschner @aparajit-pratap @sm6srw ?

Reply:
I'm fine with it as-is.